thunder/torch/__init__.py

```diff
@@ -561,12 +561,31 @@ def arange(
     return clang.arange(start=start, step=step, stop=end, device=device, dtype=dtype)
 
 
+# Infers dtype from the fill_value and dtype
+def _infer_full_dtype(fill_value: NumberLike, dtype):
+    fill_value_dtype = dtypes.numbertype_to_dtype(dtypes.to_dtype(fill_value))
+    if not dtype:
+        current_default_dtype = get_default_dtype()
+        if dtypes.is_boolean_dtype(fill_value_dtype):
+            return fill_value_dtype
+        elif dtypes.is_nonboolean_integer_dtype(fill_value_dtype):
+            return dtypes.int64
+        elif dtypes.is_complex_dtype(fill_value_dtype):
+            if current_default_dtype == torch.float64:
+                return dtypes.complex64
```

Comment:
float64 corresponds to complex128 (i.e. two doubles)
float32 -> complex64
float16 -> complex32

Reply:
Thanks for pointing that out, a slight difference is that in Torch float16->complex64
```
>>> torch.set_default_dtype(torch.float16)
>>> torch.full((2,),1.2+3j,device="cuda").dtype
torch.complex64
>>> torch.set_default_dtype(torch.float32)
>>> torch.full((2,),1.2+3j,device="cuda").dtype
torch.complex64
>>> torch.set_default_dtype(torch.float64)
>>> torch.full((2,),1.2+3j,device="cuda").dtype
torch.complex128
```

Reply:
Great detailed look, @kiya00 -- let's follow the torch mapping for now (and add a comment).

What about the bfloat16 dtype?

Reply:
> a slight difference is that in Torch float16->complex64

Nice catch, this looks like a bug in PyTorch. Internally PyTorch has the following map from Complex to Real.
https://github.com/pytorch/pytorch/blob/947a446be4360994da3b11ef01a8bdfbfe812f49/c10/core/ScalarType.h#L492-L504

Also, I think we should use this helper function as it provides the same map -
https://github.com/Lightning-AI/lightning-thunder/blob/ffbebe07bdf003c3a60f2cab88298e96b80bdbba/thunder/core/dtypes.py#L394-L395

Reply:
I updated to use Thunder helper function `corresponding_complex_dtype` and added a comment about we have this single difference than Torch: float16->complex32 (Thunder) , complex64 (Torch). I hope that's OK @mruberry